README.md
redis-clients
===============

[![Build Status](https://travis-ci.org/lykmapipo/redis-clients.svg?branch=master)](https://travis-ci.org/lykmapipo/redis-clients)
[![Dependency Status](https://img.shields.io/david/lykmapipo/redis-clients.svg?style=flat)](https://david-dm.org/lykmapipo/redis-clients)
[![npm version](https://badge.fury.io/js/redis-clients.svg)](https://badge.fury.io/js/redis-clients)

redis client factories for nodejs

## Requirements
- [Redis 2.8.0+](http://redis.io/)
- [NodeJS 8.1.4+](https://nodejs.org/en/)

## Installation
```sh
$ npm install --save redis-clients
```

## Usage

```javascript
//initialize redis-clients with default options
const redis = require('redis-clients')([options]);

//obtain normal redis client
const client = redis.client();

//set data
client.set('abc:1', 1);
client.set('abc:2', 2);
client.set('abc:3', 3);

//count keys based on pattern
redis.count('abc:*', function(error, count){
    ...
});

//count keys based on pattern
redis.count('abc:*', 'x:*', function(error, counts){
    ...
});

//obtain pub/sub clients
const { publisher, subscriber } = redis.pubsub();

//clear users data
redis.clear('users*', function(error, response){
    ...
});

//clear all data
redis.clear(function(error, response){
    ...
});

//reset and quit all clients
redis.quit();
```

## Options
- `prefix:String` - redis key prefix. default to `r`
- `separator:String` - redis key separator. default to `:`
- `redis:Object|String` - [redis](https://github.com/NodeRedis/node_redis#rediscreateclient) connections options or connection string.

To initialize `redis` with custom options use

```js
const redis = require('redis-clients')({
    prefix:'q',
    separator:'-',
    redis: {
    port: 6379,
    host: '127.0.0.1'
  }
});

...

or 

const redis = require('redis-clients')({
    prefix:'q',
    separator:'-',
    redis: 'redis://localhost:6379'
});

```

## API

### `createClient([options:Object]):Object`
Create a new instance of redis client

```js
const redis = require('redis-clients')();
const client = redis.createClient([options]);

...

```

### `pubsub():{publisher:Object, subscriber:Object}`
Obtain existing or create new instances of publisher and subscriber redis clients

```js
const redis = require('redis-clients')();
const { publisher, subscriber } = redis.pubsub();

...

```

### `client():Object`
Obtain existing or create new instance of redis client

```js
const redis = require('redis-clients')();
const client = redis.client();

...

```

### `multi():Object`
Create new instance of redis multi command object

```js
const redis = require('redis-clients')();
const client = redis.multi();

...

```

### `info(done:Function):Object`
Obtain informations of the current redis server.

```js
const redis = require('redis-clients')();
redis.info(function(error, info){
    ...
});

```

### `key(...args):String`
Build redis storage key using `prefix` and `separator` options

```js
const redis = require('redis-clients')();

//generate random key
const key = redis.key();
expect(key).to.exist;
expect(key.split(':')).to.have.length(2);

const key = redis.key('ab');
expect(key).to.be.equal('r:ab');

const key = redis.key(['users', 'ab']);
expect(key).to.be.equal('r:users:ab');

const key = redis.key('users', 'likes', 'vegetables');
expect(key).to.be.equal('r:users:likes:vegetables');

...

```

### `clear(pattern:String,[done:Function]):Array[String>`
clear data using specified pattern. if pattern not provided all data will be deleted

```js
const redis = require('redis-clients')();

//clear all data
redis.clear(function(error, responses){
    ...
});

//clear all key 'users*'
redis.clear('users',function(error, responses){
    ...
});

...

```

### `reset()`
Reset current state of redis

```js
const redis = require('redis-clients')();
redis.reset();

...

```

### `count(pattern:String, done:Function)`
Count(or obtain size) of keys based on the specified pattern. 
If no pattern specified all keys will be counted.

```js
const redis = require('redis-clients')();
redis.count('abc:*', function(error, count){
    ...
});

...

```

## Testing
* Clone this repository

* Install all development dependencies
```sh
$ npm install
```

* Then run test
```sh
$ npm test
```

## Contribute
It will be nice, if you open an issue first so that we can know what is going on, then, fork this repo and push in your ideas. Do not forget to add a bit of test(s) of what value you adding.

## Licence

The MIT License (MIT)

Copyright (c) 2017 lykmapipo & Contributors

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. 

CHANGELOG.md
#### 0.5.2 (2019-06-25)

#### 0.1.0 - 2017-09-03

* Implement client factories

test/commands.spec.js
'use strict';

/* dependencies */
const path = require('path');
const _ = require('lodash');
const { expect } = require('chai');
const faker = require('faker');

const redis = require(path.join(__dirname, '..'))();
const { set, get, hmset, hgetall } = redis.commands;


describe('commands', function () {

  before(function (done) {
    redis.clear(done);
  });

  before(function () {
    redis.reset();
  });

  describe('set', function () {

    it('should be able to set', function (done) {

      set(function (error, result) {
        expect(error).to.not.exist;
        expect(result).to.not.exist;
        done(error, result);
      });

    });

    it('should be able to set', function (done) {
      const key = faker.random.uuid();

      set(key, function (error, result) {
        expect(error).to.not.exist;
        expect(result).to.not.exist;
        done(error, result);
      });

    });

    it('should be able to set string', function (done) {
      const key = faker.random.uuid();
      const value = faker.random.word();

      set(key, value, function (error, result) {
        expect(error).to.not.exist;
        expect(result).to.exist;
        expect(result).to.be.eql(value);
        done(error, result);
      });

    });

    it('should be able to set number', function (done) {
      const key = faker.random.uuid();
      const value = faker.random.number();

      set(key, value, function (error, result) {
        expect(error).to.not.exist;
        expect(result).to.exist;
        expect(result).to.be.eql(value);
        done(error, result);
      });

    });

    it('should be able to set array', function (done) {
      const key = faker.random.uuid();
      const value = [faker.random.word(), faker.random.word()];

      set(key, value, function (error, result) {
        expect(error).to.not.exist;
        expect(result).to.exist;
        expect(result).to.be.eql(value);
        done(error, result);
      });

    });

    it('should be able to set plain object', function (done) {
      const key = faker.random.uuid();
      const value = faker.helpers.createTransaction();

      set(key, value, function (error, result) {
        expect(error).to.not.exist;
        expect(result).to.exist;
        expect(result).to.be.eql(value);
        done(error, result);
      });

    });

    it('should be able to set expiry', function (done) {
      const key = faker.random.uuid();
      const value = faker.random.word();

      set(key, value, 'EX', 1, function (error, result) {
        expect(error).to.not.exist;
        expect(result).to.exist;
        expect(result).to.be.eql(value);
        done(error, result);
      });

    });

    it('should be able to set save strategy', function (done) {
      const key = faker.random.uuid();
      const value = faker.random.word();

      set(key, value, 'EX', 1, 'NX', function (error, result) {
        expect(error).to.not.exist;
        expect(result).to.exist;
        expect(result).to.be.eql(value);
        done(error, result);
      });

    });

  });

  describe('get', function () {

    const keyString = faker.random.uuid();
    const string = faker.random.word();

    const keyNumber = faker.random.uuid();
    const number = faker.random.number();

    const keyArray = faker.random.uuid();
    const array = [faker.random.word(), faker.random.word()];

    const keyObject = faker.random.uuid();
    const object = _.omit(faker.helpers.createTransaction(), 'date');

    before(function (done) {
      set(keyString, string, done);
    });

    before(function (done) {
      set(keyNumber, number, done);
    });

    before(function (done) {
      set(keyArray, array, done);
    });

    before(function (done) {
      set(keyObject, object, done);
    });

    it('should be able to get string', function (done) {

      get(keyString, function (error, result) {
        expect(error).to.not.exist;
        expect(result).to.exist;
        expect(result).to.be.eql(string);
        done(error, result);
      });

    });

    it('should be able to get number', function (done) {

      get(keyNumber, function (error, result) {
        expect(error).to.not.exist;
        expect(result).to.exist;
        expect(result).to.be.eql(number);
        done(error, result);
      });

    });

    it('should be able to get array', function (done) {

      get(keyArray, function (error, result) {
        expect(error).to.not.exist;
        expect(result).to.exist;
        expect(result).to.be.eql(array);
        done(error, result);
      });

    });

    it('should be able to get plain object', function (done) {

      get(keyObject, function (error, result) {
        expect(error).to.not.exist;
        expect(result).to.exist;
        expect(result).to.be.eql(object);
        done(error, result);
      });

    });

  });

  describe('hmset', function () {

    it('should be able to save simple object', function (done) {
      const key = faker.random.uuid();
      const value = faker.helpers.createTransaction();

      hmset(key, value, function (error, result) {
        expect(error).to.not.exist;
        expect(result).to.exist;
        expect(result).to.be.eql(value);
        done(error, result);
      });

    });

    it('should be able to save nested object', function (done) {
      const key = faker.random.uuid();
      const value = faker.helpers.createCard();

      hmset(key, value, function (error, result) {
        expect(error).to.not.exist;
        expect(result).to.exist;
        expect(result).to.be.eql(value);
        done(error, result);
      });

    });

  });

  describe('hgetall', function () {

    const keySimpleObject = faker.random.uuid();
    const simpleObject =
      _.omit(faker.helpers.createTransaction(), 'date');

    const keyNestedObject = faker.random.uuid();
    const nestedObject =
      _.omit(faker.helpers.createCard(), 'accountHistory');

    before(function (done) {
      hmset(keySimpleObject, simpleObject, done);
    });

    before(function (done) {
      hmset(keyNestedObject, nestedObject, done);
    });

    it('should be able to get simple object', function (done) {

      hgetall(keySimpleObject, function (error, result) {
        expect(error).to.not.exist;
        expect(result).to.exist;
        expect(result).to.be.eql(simpleObject);
        done(error, result);
      });

    });

    it('should be able to get nested object', function (done) {

      hgetall(keyNestedObject, function (error, result) {
        expect(error).to.not.exist;
        expect(result).to.exist;
        expect(result).to.be.eql(nestedObject);
        done(error, result);
      });

    });

  });

  after(function (done) {
    redis.clear(done);
  });

  after(function () {
    redis.reset();
  });

});

test/redis.unit.spec.js
'use strict';

/* dependencies */
const path = require('path');
const _ = require('lodash');
const expect = require('chai').expect;
const redis = require(path.join(__dirname, '..'))();

describe('redis', function () {

  before(function (done) {
    redis.clear(done);
  });

  before(function () {
    redis.reset();
  });

  it('should be exported', function () {

    expect(redis).to.exist;

    expect(redis.init).to.exist;
    expect(redis.init).to.be.a('function');

    expect(redis.client).to.exist;
    expect(redis.client).to.be.a('function');

    expect(redis.pubsub).to.exist;
    expect(redis.pubsub).to.be.a('function');

    expect(redis._client).to.not.exist;
    expect(redis.publisher).to.not.exist;
    expect(redis.subscriber).to.not.exist;

  });

  describe('init & client & pubsub client', function () {
    before(function () {
      redis.reset();
    });

    it('should return same redis instances once initialized', function () {
      expect(redis.client()._id)
        .to.be.equal(redis.client()._id);

      expect(redis.pubsub().publisher._id)
        .to.be.equal(redis.pubsub().publisher._id);

      expect(redis.pubsub().subscriber._id)
        .to.be.equal(redis.pubsub().subscriber._id);
    });

    it('should return same redis instances once initialized', function () {
      //require redis
      const rediz = require(path.join(__dirname, '..'))();

      expect(rediz.client()._id)
        .to.be.equal(redis.client()._id);

      expect(rediz.pubsub().publisher._id)
        .to.be.equal(redis.pubsub().publisher._id);

      expect(rediz.pubsub().subscriber._id)
        .to.be.equal(redis.pubsub().subscriber._id);
    });

    after(function () {
      redis.reset();
    });

  });


  describe('options', function () {
    beforeEach(function () {
      redis.reset();
    });

    it('should have default options', function () {
      expect(redis.defaults).to.exist;
      expect(redis.defaults).to.be.an('object');
      expect(redis.defaults).to.be.eql({
        prefix: 'r',
        separator: ':',
        redis: {
          port: 6379,
          host: '127.0.0.1'
        }
      });
    });

    it('should be able to set key prefix options', function () {

      redis.defaults.prefix = 'q';

      expect(redis.defaults).to.exist;
      expect(redis.defaults).to.be.an('object');
      expect(redis.defaults).to.be.eql({
        prefix: 'q',
        separator: ':',
        redis: {
          port: 6379,
          host: '127.0.0.1'
        }
      });

    });

    it('should be able to set key separator options', function () {

      redis.defaults.separator = '-';

      expect(redis.defaults).to.exist;
      expect(redis.defaults).to.be.an('object');
      expect(redis.defaults).to.be.eql({
        prefix: 'r',
        separator: '-',
        redis: {
          port: 6379,
          host: '127.0.0.1'
        }
      });

    });

    after(function () {
      redis.reset();
    });

  });

  it('should be able to initialize multi object', function () {
    const multi = redis.multi();
    expect(multi).to.exist;
    expect(multi.exec).to.exist;
    expect(multi.exec).to.be.a('function');
  });

  it('should be able to generate a key', function () {
    const key = redis.key();
    expect(key).to.exist;
    expect(key.split(':')).to.have.length(2);
  });

  it('should be able to create redis key', function () {
    const key1 = redis.key('ab');
    expect(key1).to.be.equal('r:ab');

    const key2 = redis.key(['users', 'ab']);
    expect(key2).to.be.equal('r:users:ab');

    const key3 = redis.key('users', 'likes', 'vegetables');
    expect(key3).to.be.equal('r:users:likes:vegetables');
  });

  it('should be able to create a new instance of redis client', function () {
    const client = redis.createClient();
    expect(client).to.exist;
  });

  it('should be able to instantiate redis client', function () {
    const client = redis.client();
    expect(client).to.exist;
  });

  it('should be able to instantiate pub/sub redis clients', function () {
    const { publisher, subscriber } = redis.pubsub();
    expect(publisher).to.exist;
    expect(subscriber).to.exist;
  });

  it('should be able to connect to redis', function () {
    redis.init();

    expect(redis._client).to.exist;
    expect(redis._client).to.not.be.null;

    expect(redis.publisher).to.exist;
    expect(redis.publisher).to.not.be.null;

    expect(redis.subscriber).to.exist;
    expect(redis.subscriber).to.not.be.null;

  });

  it('should be able to reset it state', function () {
    redis.reset();

    expect(redis._client).to.be.null;
    expect(redis.publisher).to.be.null;
    expect(redis.subscriber).to.be.null;

  });

  it('should be able to obtain redis server info', function (done) {
    redis.info(function (error, info) {
      expect(error).to.not.exist;
      expect(info).to.exist;
      done(error, info);
    });
  });


  describe('clear', function () {
    beforeEach(function (done) {
      redis.clear(done);
    });

    it('should be able to clear all data', function (done) {
      const key = redis.key();
      redis.client().set(key, 1, function (error) {
        if (error) {
          done(error);
        } else {
          redis.clear(function (error, response) {
            expect(error).to.not.exist;
            expect(response).to.have.length(1);
            expect(response).to.contain.members([1]);
            done(error, response);
          });
        }
      });
    });

    it('should be able to clear data with given patterns', function (
      done) {
      const key = redis.key('users');
      redis.client().set(key, 1, function (error) {
        if (error) {
          done(error);
        } else {
          redis.clear('users', function (error, response) {
            expect(error).to.not.exist;
            expect(response).to.have.length(1);
            expect(response).to.contain.members([1]);
            done(error, response);
          });
        }
      });
    });
  });

  describe('count', function () {

    before(function (done) {
      const client = redis.client();
      client.multi()
        .set('abc:1', 1)
        .set('abc:2', 2)
        .set('abc:3', 3)
        .exec(done);
    });

    it(
      'should be able to count size of keys of specifed pattern',
      function (done) {

        redis
          .count('abc:*', function (error, count) {
            expect(error).to.not.exist;
            expect(count).to.exist;
            expect(count).to.be.equal(3);
            done(error, count);
          });

      });

    it(
      'should be able to count size of keys of specifed patterns',
      function (done) {

        redis
          .count('abc:*', 'xx:*', function (error, counts) {
            expect(error).to.not.exist;
            expect(counts).to.exist;
            expect(counts).to.have.length(2);
            expect(_.first(counts)).to.be.equal(3);
            expect(_.last(counts)).to.be.equal(0);
            done(error, counts);
          });

      });

  });

  describe('connetion init', function () {

    before(function () {
      redis.reset();
    });

    it('should support redis connection string', function () {
      const options = { redis: 'redis://localhost:6379' };
      const client = redis.createClient(options);
      expect(client).to.exist;
    });

    after(function () {
      redis.reset();
    });

  });


  after(function (done) {
    redis.clear(done);
  });

  after(function () {
    redis.reset();
  });

});

src/commands.js
'use strict';

/**
 * @name commands
 * @description redis comman helper
 * @author lally elias<lallyelias87@gmail.com>
 * @since 0.4.0
 * @version 0.1.0
 * @singleton
 * @type {Object}
 */


/* dependencies */
const path = require('path');
const _ = require('lodash');
const flat = require('flat');
const unflat = require('flat').unflatten;
const redis = require(path.join(__dirname, 'redis'));


/* noop */
const noop = function () {};


/* stringify data to save to redis */
const stringify = function (value) {
  /* clone */
  let _value = _.clone(value);
  try {
    if (!_.isString(value)) {
      _value = JSON.stringify(value);
    }
    return _value;
  } catch (error) {
    return _value;
  }
};

/* parse data from redis */
const parse = function (value) {
  /* clone */
  let _value = _.clone(value);
  try {
    _value = JSON.parse(_value);
    return _value;
  } catch (error) {
    return _value;
  }
};


/**
 * @name set
 * @function set
 * @description Set key to hold the value. 
 * If key already holds a value, it is overwritten, 
 * regardless of its type.
 * 
 * @param {String} key key
 * @param {Mixed} value value
 * @param {String} [expiry] expiry strategy(i.e PX or EX)
 * @param {Number} [time] expiry time(i.e seconds or milliseconds)
 * @param {String} [strategy] save strategy(i.e NX or XX)
 * @param {Function} [done] a callback to invoke on success or failure
 * @see {@link https://redis.io/commands/set|SET}
 * @since 0.4.0
 * @version 0.1.0
 * @public
 */
exports.set = function set(key, value, expiry, time, strategy, done) {

  /* do nothing */
  if (_.isFunction(key)) {
    return (key && key());
  }

  /* do nothing */
  if (_.isFunction(value)) {
    return (value && value());
  }

  /* ensure client */
  const client = redis.client();

  /* prepare */
  const _key = redis.key(key);
  const _value = stringify(value);
  const _expiry = expiry && _.isString(expiry) ? expiry : undefined;
  const _time = time && _.isNumber(time) ? time : undefined;
  const _strategy = strategy && _.isString(strategy) ? strategy : undefined;


  /* callback */
  let _cb = noop;
  _cb = _.isFunction(expiry) ? expiry : _cb;
  _cb = _.isFunction(time) ? time : _cb;
  _cb = _.isFunction(strategy) ? strategy : _cb;
  _cb = _.isFunction(done) ? done : _cb;
  const cb = function (error) {
    return _cb && _cb(error, value);
  };


  /* compact */
  const args =
    _.compact([_key, _value, _expiry, _time, _strategy, cb]);


  /* set */
  return client.set.call(client, ...args);

};


/**
 * @name get
 * @function get
 * @description Get the value of key
 * 
 * @param {String} key key
 * @param {Function} [done] a callback to invoke on success or failure
 * @see {@link https://redis.io/commands/get|GET}
 * @since 0.4.0
 * @version 0.1.0
 * @public
 */
exports.get = function get(key, done) {

  /* do nothing */
  if (_.isFunction(key)) {
    return (key && key());
  }

  /* ensure client */
  const client = redis.client();

  /* prepare */
  const _key = redis.key(key);

  /* callback */
  let _cb = noop;
  _cb = _.isFunction(done) ? done : _cb;
  const cb = function (error, value) {
    return _cb && _cb(error, parse(value));
  };


  /* compact */
  const args = _.compact([_key, cb]);


  /* set */
  return client.get.call(client, ...args);

};


/**
 * @name hmset
 * @function hmset
 * @description Sets the specified fields to their respective values 
 * in the hash stored at key. 
 * 
 * @param {String} key key
 * @param {Object} value value
 * @param {Function} [done] a callback to invoke on success or failure
 * @see {@link https://redis.io/commands/hmset|HMSET}
 * @since 0.4.0
 * @version 0.1.0
 * @public
 */
exports.hmset = function hmset(key, value, done) {

  /* do nothing */
  if (_.isFunction(key)) {
    return (key && key());
  }

  /* do nothing */
  if (_.isFunction(value)) {
    return (value && value());
  }

  /* ensure client */
  const client = redis.client();

  /* prepare */
  const _key = redis.key(key);
  const _value = flat(value);

  /* callback */
  let _cb = noop;
  _cb = _.isFunction(done) ? done : _cb;
  const cb = function (error) {
    return _cb && _cb(error, value);
  };

  /* compact */
  const args = _.compact([_key, _value, cb]);


  /* set */
  return client.hmset.call(client, ...args);

};


/**
 * @name hgetall
 * @function hgetall
 * @description Returns all fields and values of the hash stored at key
 * 
 * @param {String} key key
 * @param {Function} [done] a callback to invoke on success or failure
 * @see {@link https://redis.io/commands/hgetall|HGETALL}
 * @since 0.4.0
 * @version 0.1.0
 * @public
 */
exports.hgetall = function hgetall(key, done) {

  /* do nothing */
  if (_.isFunction(key)) {
    return (key && key());
  }

  /* ensure client */
  const client = redis.client();

  /* prepare */
  const _key = redis.key(key);

  /* callback */
  let _cb = noop;
  _cb = _.isFunction(done) ? done : _cb;
  const cb = function (error, value) {
    return _cb && _cb(error, unflat(value));
  };


  /* compact */
  const args = _.compact([_key, cb]);


  /* set */
  return client.hgetall.call(client, ...args);

};

src/redis.js
'use strict';


/**
 * @module redis-clients
 * @name redis-clients
 * @description redis client factories
 * @author lally elias<lallyelias87@gmail.com>
 * @see {@link https://github.com/lykmapipo/redis-clients}
 * @since 0.1.0
 * @version 0.1.0
 * @singleton
 * @public
 * @type {Object}
 */


//global dependencies
const url = require('url');
const _ = require('lodash');
const redis = require('redis');
const uuid = require('uuid');


//local initializations
const noop = function () {};


//reference to all created redis clients
//mainly used for safe shutdown and resource cleanups
exports._clients = [];


//defaults settings
const defaults = {
  prefix: 'r',
  separator: ':',
  redis: {
    port: 6379,
    host: '127.0.0.1'
  }
};


/**
 * @name defaults
 * @description default redis client connection options
 * @type {Object}
 * @since 0.1.0
 * @public
 */
exports.defaults = _.merge({}, defaults);


/**
 * @function
 * @name createClient
 * @description instantiate new redis client if not exists
 * @return {Object} an instance of redis client
 * @since 0.1.0
 * @public
 */
exports.createClient = function (options) {

  //merge options
  options = _.merge({}, exports.defaults, options);

  //support connection string url
  if (_.isString(options.redis)) {

    // parse the url
    const connection = url.parse(options.redis, true /* parse query string */ );
    if (connection.protocol !== 'redis:') {
      throw new Error('Invalid Connection String. Require redis: protocol');
    }

    options.redis = {
      port: connection.port || 6379,
      host: connection.hostname,
      db: (connection.pathname ? connection.pathname.substr(1) : null) ||
        connection.query.db || 0,
      // see https://github.com/mranney/node_redis#rediscreateclient
      options: connection.query
    };

    if (connection.auth) {
      options.redis.auth = connection.auth.replace(/.*?:/, '');
    }

  }

  //instantiate a redis client
  const socket = options.redis.socket;
  const port = !socket ? (options.redis.port || 6379) : null;
  const host = !socket ? (options.redis.host || '127.0.0.1') : null;
  const client =
    redis.createClient(socket || port, host, options.redis.options);

  //authenticate
  if (options.redis.auth) {
    client.auth(options.redis.auth);
  }

  //select database
  if (options.redis.db) {
    client.select(options.redis.db);
  }

  //remember created client(s) for later safe shutdown
  exports._clients = _.compact([].concat(exports._clients).concat(client));

  return client;

};


/**
 * @function
 * @name init
 * @description initialize redis client and pubsub
 * @return {Object} redis client
 * @since 0.1.0
 * @public
 */
exports.init = exports.client = function () {

  //initialize normal client
  if (!exports._client) {
    exports._client = exports.createClient();
    exports._client._id = Date.now();
  }

  //initialize publisher and subscriber clients
  exports.pubsub();

  //return a normal redis client
  return exports._client;

};


/**
 * @function
 * @name pubsub
 * @description instantiate redis pub-sub clients pair if not exists
 * @since 0.1.0
 * @public
 */
exports.pubsub = function () {

  //create publisher if not exists
  if (!exports.publisher) {
    exports.publisher = exports.createClient();
    exports.publisher._id = Date.now();
  }

  //create subscriber if not exist
  if (!exports.subscriber) {
    exports.subscriber = exports.createClient();
    exports.subscriber._id = Date.now();
  }

  //exports pub/sub clients
  return { publisher: exports.publisher, subscriber: exports.subscriber };
};


/**
 * @function
 * @name multi
 * @description initialize redis multi command object
 * @return {Object} redis multi command object
 * @since 0.3.0
 * @see {@link https://github.com/NodeRedis/node_redis#clientmulticommands}
 * @public
 */
exports.multi = function () {
  //ensure clients
  const client = exports.client();

  //obtain client
  const multi = client.multi();

  return multi;
};


/**
 * @function
 * @name info
 * @description collect redis server health information
 * @param  {Function} done a callback to invoke on success or failure
 * @return {Object}        server details
 * @since 0.1.0
 * @public
 */
exports.info = function (done) {

  exports.client().info(function (error /*, info*/ ) {
    // jshint camelcase:false
    done(error, exports._client.server_info);
    // jshint camelcase:true
  });

};


/**
 * @function
 * @name key
 * @description prepare data storage key
 * @param  {String|String[]} key valid data store key
 * @since 0.1.0
 * @public
 */
exports.key = function (...args) {

  //concatenate key is varargs
  let key = [].concat(...args);

  //ensure key
  if (key.length === 0) {
    key = key.concat(uuid.v1());
  }

  key = [exports.defaults.prefix].concat(key);

  //join key using separator
  key = key.join(exports.defaults.separator);

  return key;

};


/**
 * @function
 * @name count
 * @description count the number of keys that match specified pattern
 * @param {String|Array} [patterns] single or collection of patterns
 * @param {Function} done a callback to invoke on success or failure
 * @since 0.2.0
 * @public
 * @return {Number|Array} count per specified pattern in that order respectively
 * @see {@link https://redis.io/commands/eval}
 * @see {@link http://maaxiim.blogspot.ru/2012/09/implementing-redis-count-command-in-lua.html}
 */
exports.count = exports.size = function (...patterns) {

  //normalize patterns to array
  patterns = [].concat(...patterns);

  //compact and ensure unique patterns
  patterns = _.uniq(_.compact(patterns));

  //obtain callback
  const done = _.last(patterns);

  //drop callback if provided
  if (_.isFunction(done)) {
    patterns = _.initial(patterns);
  }

  //ensure patterns to count
  if (patterns && patterns.length > 0) {

    //get a client
    const client = exports.multi();

    //count for each pattern

    _.forEach(patterns, function (pattern) {

      //prepare count LUA script per pattern
      const script =
        (['return #redis.pcall("keys", "', pattern, '")'].join(''));

      //count using a lua script
      client.eval(script, 0);

    });

    client.exec(function (error, count) {

      //normalize count
      if (count) {
        count = count.length > 1 ? count : _.first(count);
      }

      done(error, count);

    });
  }

  //reply with bad request
  //as no collections specified
  else {
    let error = new Error('Missing Count Patterns');
    error.status = 400;
    done(error);
  }


};


/**
 * @function
 * @name reset
 * @description quit and reset redis clients states
 * @since 0.1.0
 * @public
 */
exports.reset = exports.quit = function () {

  //clear subscriptions and listeners
  if (exports.subscriber) {
    exports.subscriber.unsubscribe();
    exports.subscriber.removeAllListeners();
  }

  //quit all clients
  _.forEach(exports._clients, function (_client) {
    //TODO do they shutdown immediately
    //TODO check kue how they handle this
    _client.quit();
  });

  //TODO is there any memory leaks here?
  //TODO notify others before reset the clients for better cleanup
  //reset clients
  exports._client = null;
  exports.publisher = null;
  exports.subscriber = null;

  //reset settings
  exports.defaults = _.merge({}, defaults);

  //reset clients
  exports._clients = [];

};


/**
 * @function
 * @name clear
 * @description clear all data saved and their key
 * @param {String} [pattern] pattern of keys to be removed
 * @param {Function} done a callback to invoke on success or failure
 * @since 0.1.0
 * @public
 */
exports.clear = function (pattern, done) {

  //normalize arguments
  if (pattern && _.isFunction(pattern)) {
    done = pattern;
    pattern = undefined;
  }

  //ensure callback
  if (!done) {
    done = noop;
  }

  //prepare clear all key regex
  pattern = _.compact([exports.defaults.prefix, pattern]);
  if (pattern.length > 1) {
    pattern = pattern.join(exports.defaults.separator);
  }
  pattern = [pattern].concat(['*']).join('');

  //clear data in transaction
  exports.client().keys(pattern, function (error, keys) {

    //back-off in case there is error
    if (error) {
      done(error);
    }

    //execute delete using redis multi command
    else {

      //initiate multi to run all commands atomically
      const _client = exports.multi();

      //queue commands
      _.forEach(keys, function (key) {
        _client.del(key);
      });

      //execute commands
      _client.exec(done);
    }

  });

};